### PR TITLE
LinkageProblemCauseAnnotator to take ClassPathBuilder to configure repositories

### LINKED ISSUE #1609: NullPointerException when annotating com.google.cloud:spring-cloud-gcp-core:jar:2.0.0-SNAPSHOT
When I tried to apply LinkageChecker to spring-cloud-gcp, it threw a NullPointerException for dependency graph of `com.google.cloud:spring-cloud-gcp-core:jar:2.0.0-SNAPSHOT`.

```
Caused by: java.lang.NullPointerException
    at com.google.cloud.tools.opensource.dependencies.DependencyGraph.levelOrder (DependencyGraph.java:259)
    at com.google.cloud.tools.opensource.dependencies.DependencyGraph.from (DependencyGraph.java:244)
    at com.google.cloud.tools.opensource.dependencies.DependencyGraphBuilder.buildDependencyGraph (DependencyGraphBuilder.java:204)
    at com.google.cloud.tools.opensource.dependencies.DependencyGraphBuilder.buildMavenDependencyGraph (DependencyGraphBuilder.java:193)
    at com.google.cloud.tools.opensource.classpath.ClassPathBuilder.resolveWithMaven (ClassPathBuilder.java:80)
    at com.google.cloud.tools.opensource.classpath.LinkageProblemCauseAnnotator.annotate (LinkageProblemCauseAnnotator.java:51)
    at com.google.cloud.tools.dependencies.enforcer.LinkageCheckerRule.execute (LinkageCheckerRule.java:216)
    at org.apache.maven.plugins.enforcer.EnforceMojo.execute (EnforceMojo.java:205)
```

Screen Shot 2020-08-19 at 10 38 35

Screen Shot 2020-08-19 at 10 41 48

The artifact is locally available:

```
suztomo-macbookpro44% head  ~/.m2/repository/com/google/cloud/spring-cloud-gcp-core/2.0.0-SNAPSHOT/spring-cloud-gcp-core-2.0.0-SNAPSHOT.pom
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
		 xmlns="http://maven.apache.org/POM/4.0.0"
		 xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
	<modelVersion>4.0.0</modelVersion>
	<parent>
		<artifactId>spring-cloud-gcp</artifactId>
		<groupId>com.google.cloud</groupId>
		<version>2.0.0-SNAPSHOT</version>
	</parent>
```

# Diagnosis

Why `com.google.cloud:spring-cloud-gcp-core:jar:2.0.0-SNAPSHOT` is not resolvable? It says `org.apache.maven.model.resolution.UnresolvableModelException: Could not find artifact io.projectreactor:reactor-bom:pom:2020.0.0-M2 in  (https://repo1.maven.org/maven2/)`.

Screen Shot 2020-08-19 at 10 54 20

The version does not exist for the POM artifact in the Maven Central. https://search.maven.org/artifact/io.projectreactor/reactor-bom. Where is it getting then? Repository element somewhere?

It's in spring-milestone repository

```
[INFO] Downloading from spring-milestones: https://repo.spring.io/milestone/io/projectreactor/reactor-bom/2020.0.0-M2/reactor-bom-2020.0.0-M2.pom
```

The cause is `ClassPathBuilder.resolveWithMaven` does not take `repository` section of the Maven project into account.

### PR DESCRIPTION
**Pull request by suztomo**
Fixes #1609

### CODE REVIEW

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageProblemCauseAnnotator.java [hunk 1]
```diff
@@ -47,7 +52,6 @@ public static void annotate(ClassPathResult rootResult, Iterable<LinkageProblem>
       ClassPathResult subtreeResult = cache.get(sourceArtifact);
       if (subtreeResult == null) {
         // Resolves the dependency graph with the source artifact at the root.
-        ClassPathBuilder classPathBuilder = new ClassPathBuilder();
         subtreeResult = classPathBuilder.resolveWithMaven(sourceArtifact);
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageProblemCauseAnnotator.java:55]: classPathBuilder looks like it can be null here.

Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageProblemCauseAnnotator.java:55]: classPathBuilder parameter cannot be null; otherwise it throws NullPointerException here.

Initially I thought making the parameter `@Nullable`. However, the enforcer rule and LinkageCheckerMain both need to resolve dependency graphs using their ClassPathBuilder.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java [hunk 2]
```diff
@@ -100,7 +101,7 @@ public static void main(String[] arguments)
         }
 
         if (classPathResult != null) {
```
Comment (suztomo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java:103]: if `classPathResult` is not null, then `classPathBuilder` is not null.

Screen Shot 2020-09-10 at 13 37 36

**Review by suztomo — COMMENTED** [review]
@elharo PTAL.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java [hunk 3]
```diff
@@ -67,7 +67,8 @@ public static void main(String[] arguments)
         List<ArtifactProblem> artifactProblems = new ArrayList<>();
         // classPathResult is kept null if JAR files are specified in the argument
         ClassPathResult classPathResult = null;
-    
+        ClassPathBuilder classPathBuilder = null;
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java:70]: using these in two variables in only one branch of the if-else block is a bit of a code smell. I wonder if there's an alternative approach that doesn't use nullness as flags?

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageProblemCauseAnnotator.java (hunk 1 above)
Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageProblemCauseAnnotator.java:55]: It is a public method so anything can be passed. If you expect to be not null, it's worth checking.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java (hunk 3 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java:71]: It now relies on `artifactsInArguments` variable. It's empty when we specify JAR files.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageProblemCauseAnnotator.java (hunk 1 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageProblemCauseAnnotator.java:59]: Added `checkNotNull()`s.

**Review by suztomo — COMMENTED** [review]
@elharo PTAL.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java (hunk 3 above)
Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java:71]: This is becoming more complex, not less. I think the fundamental issue is that two very different cases are being managed with lots of if-else instead of different methods or objects. When I looked at this code in my IDE, it was quite hard to follow.

Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java:71]: let me try to split the 2 (3?) cases into methods.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java [hunk 4]
```diff
@@ -57,85 +55,19 @@ public static void main(String[] arguments)
...
-        }
+      if (linkageCheckerArguments.hasInput()) {
+        // artifactsInArguments is not empty if a BOM or artifacts are specified in the argument.
+        // If JAR files are specified, it's empty.
+        ImmutableList<Artifact> artifactsInArguments = linkageCheckerArguments.getArtifacts();
 
+        ImmutableSet<LinkageProblem> linkageProblems =
+            artifactsInArguments.isEmpty()
+                ? runWithJarFiles(linkageCheckerArguments)
+                : runWithArtifacts(linkageCheckerArguments);
```
Comment (suztomo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java:66]: Now the the case of JAR file and the case of artifacts are separated.

**Review by suztomo — COMMENTED** [review]
@elharo PTAL.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java [hunk 5]
```diff
@@ -145,4 +77,102 @@ public static void main(String[] arguments)
...
+  private static ImmutableSet<LinkageProblem> runWithArtifacts(
+      LinkageCheckerArguments linkageCheckerArguments)
+      throws IOException, RepositoryException, TransformerException, XMLStreamException {
+    ImmutableList<Artifact> artifactsInArguments = linkageCheckerArguments.getArtifacts();
+
+    // When a BOM or Maven artifacts are passed as arguments, resolve the dependencies.
+    DependencyGraphBuilder dependencyGraphBuilder =
+        new DependencyGraphBuilder(linkageCheckerArguments.getMavenRepositoryUrls());
+    ClassPathBuilder classPathBuilder = new ClassPathBuilder(dependencyGraphBuilder);
+    ClassPathResult classPathResult = classPathBuilder.resolve(artifactsInArguments, false);
```
Comment (suztomo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java:124]: classPathBuilder is always non-null.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java [hunk 6]
```diff
@@ -145,4 +77,102 @@ public static void main(String[] arguments)
       System.err.println(ex.getMessage());
     }
   }
+
+  private static ImmutableSet<LinkageProblem> runWithJarFiles(
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java:81]: checkJarFiles?

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java [hunk 7]
```diff
@@ -145,4 +77,102 @@ public static void main(String[] arguments)
...
+    }
+
+    if (!linkageProblems.isEmpty()) {
+      System.out.println(LinkageProblem.formatLinkageProblems(linkageProblems));
+    }
+
+    return linkageProblems;
+  }
+
+  private static ImmutableSet<LinkageProblem> runWithArtifacts(
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java:115]: checkArtifacts?

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java [hunk 8]
```diff
@@ -145,4 +77,102 @@ public static void main(String[] arguments)
...
+      ClassReferenceGraph graph = linkageChecker.getClassReferenceGraph();
+      linkageProblems =
+          linkageProblems.stream()
+              .filter(
+                  (LinkageProblem problem) ->
+                      graph.isReachable(problem.getSourceClass().getBinaryName()))
+              .collect(toImmutableSet());
+    }
+
+    Path writeAsExclusionFile = linkageCheckerArguments.getOutputExclusionFile();
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java:101]: This chunk of code could be extracted to a `writeExclusionFile` helper method to avoid duplication

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java [hunk 9]
```diff
@@ -145,4 +77,102 @@ public static void main(String[] arguments)
...
+    ImmutableSet<ClassPathEntry> entryPoints =
+        ImmutableSet.copyOf(inputClassPath.subList(0, artifactsInArguments.size()));
+
+    LinkageChecker linkageChecker =
+        LinkageChecker.create(
+            inputClassPath, entryPoints, linkageCheckerArguments.getInputExclusionFile());
+    ImmutableSet<LinkageProblem> linkageProblems = linkageChecker.findLinkageProblems();
+
+    if (linkageCheckerArguments.getReportOnlyReachable()) {
+      ClassReferenceGraph graph = linkageChecker.getClassReferenceGraph();
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java:137]: the body of this if block can probably be pulled out to a helper method (or maybe the whole if block) to avoid the duplication

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java [hunk 10]
```diff
@@ -145,4 +77,102 @@ public static void main(String[] arguments)
...
+    LinkageProblemCauseAnnotator.annotate(classPathBuilder, classPathResult, linkageProblems);
+
+    Path writeAsExclusionFile = linkageCheckerArguments.getOutputExclusionFile();
+    if (writeAsExclusionFile != null) {
+      ExclusionFiles.write(writeAsExclusionFile, linkageProblems);
+      System.out.println("Wrote the linkage errors as exclusion file: " + writeAsExclusionFile);
+      return ImmutableSet.of();
+    }
+
+    if (!linkageProblems.isEmpty()) {
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java:155]: tempted to move this if block to a printLinkageProblems method too

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java (hunk 6 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java]: Updated.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java (hunk 7 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java]: Updated

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java (hunk 8 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java:127]: Updated.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java (hunk 9 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java]: Done.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java (hunk 10 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java:165]: I searched the usage of `formatLinkageProblems`.  The enforcer rule uses logger. That `printLinkageProblems` method would be only used by LinkageCheckerMain. So it's better to have if statement here.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java [hunk 11]
```diff
@@ -145,4 +77,112 @@ public static void main(String[] arguments)
...
+              .collect(toImmutableSet());
+    }
+    return linkageProblems;
+  }
+
+  /**
+   * Writes {@code linkageProblems} to a file and returns true if {@code linkageCheckerArguments}
+   * specifies the option; otherwise returns false.
+   */
+  private static boolean writeExclusionFile(
```
Comment (suztomo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java:124]: @elharo Do you think of a better name for such function that does nothing depending on the argument? I was thinking `writeExclusionFileIfnecessary` but this sounds redundant.

**Comment by suztomo** [comment]
@elharo PTAL

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java [hunk 12]
```diff
@@ -88,6 +88,23 @@ public static void main(String[] arguments)
...
+      System.out.println(LinkageProblem.formatLinkageProblems(linkageProblems));
+    }
+
+    return linkageProblems;
+  }
+
+  private static ImmutableSet<LinkageProblem> filterReachable(
+      LinkageCheckerArguments linkageCheckerArguments,
+      LinkageChecker linkageChecker,
+      ImmutableSet<LinkageProblem> linkageProblems) {
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java:107]: I don't think you need this linkageProblems argument

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java (hunk 11 above)
Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java:124]: writeExclusionFile is fine

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java [hunk 13]
```diff
@@ -97,22 +114,26 @@ public static void main(String[] arguments)
...
+  }
 
+  /**
+   * Writes {@code linkageProblems} to a file and returns true if {@code linkageCheckerArguments}
+   * specifies the option; otherwise returns false.
+   */
+  private static boolean writeExclusionFile(
+      LinkageCheckerArguments linkageCheckerArguments, ImmutableSet<LinkageProblem> linkageProblems)
+      throws TransformerException, XMLStreamException, IOException {
     Path writeAsExclusionFile = linkageCheckerArguments.getOutputExclusionFile();
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java:127]: exclusionFilePath or just path

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java [hunk 14]
```diff
@@ -88,6 +88,23 @@ public static void main(String[] arguments)
...
+    }
+
+    return linkageProblems;
+  }
+
+  private static ImmutableSet<LinkageProblem> filterReachable(
+      LinkageCheckerArguments linkageCheckerArguments,
+      LinkageChecker linkageChecker,
+      ImmutableSet<LinkageProblem> linkageProblems) {
     if (linkageCheckerArguments.getReportOnlyReachable()) {
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java:108]: I'm tempted to pull the if out of this method and not pass the linkageChecker arguments in

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java [hunk 15]
```diff
@@ -57,85 +55,19 @@ public static void main(String[] arguments)
...
-          for (LinkageProblem linkageProblem : linkageProblems) {
-            ClassFile targetClass = linkageProblem.getTargetClass();
-            if (targetClass != null) {
-              problematicJars.add(targetClass.getClassPathEntry());
-            }
-            ClassFile sourceClassFile = linkageProblem.getSourceClass();
-            problematicJars.add(sourceClassFile.getClassPathEntry());
-          }
-          System.out.println(classPathResult.formatDependencyPaths(problematicJars.build()));
-        }
```
Comment (suztomo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java:128]: This logic has moved to `formatDependencyPaths`.

File: enforcer-rules/src/it/bom-project-error/verify.groovy [hunk 16]
```diff
@@ -28,7 +28,7 @@ Class org.apache.log.Logger is not found;
       unselected: io.grpc:grpc-core:1.17.1 (compile) / com.google.guava:guava:26.0-android (compile)
 ''')
 
-assert buildLog.contains('''[ERROR] Problematic artifacts in the dependency tree:
+assert buildLog.contains('''Problematic artifacts in the dependency tree:
```
Comment (suztomo) [inline enforcer-rules/src/it/bom-project-error/verify.groovy:31]: Because linkage problems and the dependency paths are printed as one message, this no longer carry `[ERROR] `.

File: enforcer-rules/src/it/fail-build-for-linkage-errors/verify.groovy [hunk 17]
```diff
@@ -20,7 +20,7 @@ assert buildLog.contains('''\
 ''')
 
 assert buildLog.contains('''\
-[ERROR] Problematic artifacts in the dependency tree:
+Problematic artifacts in the dependency tree:
```
Comment (suztomo) [inline enforcer-rules/src/it/fail-build-for-linkage-errors/verify.groovy:23]: Because linkage problems and the dependency paths are printed as one message, this no longer carry `[ERROR] `.

File: enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java [hunk 18]
```diff
@@ -385,20 +382,4 @@ private ClassPathResult findBomClasspath(
     }
     return result;
   }
-
-  private String dependencyPathsOfProblematicJars(
```
Comment (suztomo) [inline enforcer-rules/src/main/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRule.java:389]: This has moved to formatLinkageProblems.

File: enforcer-rules/src/test/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRuleTest.java [hunk 19]
```diff
@@ -232,12 +231,12 @@ public void testExecute_shouldFailForBadProject() throws RepositoryException {
     } catch (EnforcerRuleException ex) {
       // pass
       ArgumentCaptor<String> errorMessageCaptor = ArgumentCaptor.forClass(String.class);
-      verify(mockLog, times(2)).error(errorMessageCaptor.capture());
+      verify(mockLog, times(1)).error(errorMessageCaptor.capture());
```
Comment (suztomo) [inline enforcer-rules/src/test/java/com/google/cloud/tools/dependencies/enforcer/LinkageCheckerRuleTest.java:234]: Now one `error()` prints both linkage problems and dependency paths.

**Review by suztomo — COMMENTED** [review]
@elharo PTAL

**Review by elharo — APPROVED** [review]
